ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * {@link Encryptor} implementation for encrypting/decrypting PASSWORD type
+ * properties
+ */
+
+@Singleton
+public class PasswordPropertiesEncryptor implements Encryptor<Map<String, String>> {
+
+  private static final String ENCRYPTED_PROPERTY_PREFIX = "${enc=aes256_hex, value=";
+  private static final String ENCRYPTED_PROPERTY_SCHEME = ENCRYPTED_PROPERTY_PREFIX + "%s}";
+
+  private final EncryptionService encryptionService;
+  private final Map<Long, Map<String, Set<String>>> clusterPasswordProperties = new ConcurrentHashMap<>();
+
+  @Inject
+  public PasswordPropertiesEncryptor(EncryptionService encryptionService, AmbariEventPublisher ambariEventPublisher) {
+    this.encryptionService = encryptionService;
+    ambariEventPublisher.register(this);
+  }
+
+  /**
+   * It is a must that any caller of this method adds the following additional
+   * information on the following indexes in <code>additionalInfo</code> with the
+   * following types:
+   * <ul>
+   * <li>[0] - <code>org.apache.ambari.server.state.Cluster</code> (the cluster
+   * whose properties should be scanned for PASSWORD properties)</li>
+   * <li>[1] - <code>java.lang.String</code> (to keep properties only that match
+   * this configuration type )</li>
+   * </ul>
+   */
+  @Override
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaInfo) {
+    try {
+      final Map<String, String> encryptedProperties = new HashMap<>(properties.size());
+      if (properties != null) {
+        final Cluster cluster = (Cluster) additionaInfo[0];
+        final String configType = (String) additionaInfo[1];
+        final Set<String> passwordProperties = getPasswordProperties(cluster, configType);
+        if (CollectionUtils.isNotEmpty(passwordProperties)) {
+          String encryptedValue;
+          for (Map.Entry<String, String> property : properties.entrySet()) {
+            encryptedValue = passwordProperties.contains(property.getKey()) && !isEncryptedPassword(property.getValue())
+                ? encryptAndDecoratePropertyValue(property.getValue())
+                : property.getValue();
+            encryptedProperties.put(property.getKey(), encryptedValue);
+          }
+        }
+      }
+      return encryptedProperties;
+    } catch (Exception e) {
+      throw new AmbariRuntimeException("Error while encrypting sensitive data", e);
+    }
+  }
+
+  private boolean isEncryptedPassword(String password) {
+    return password != null && password.startsWith(ENCRYPTED_PROPERTY_PREFIX); // assuming previous encryption by this class
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, String configType) throws AmbariException {
+    final long clusterId = cluster.getClusterId();
+    if (!clusterPasswordProperties.containsKey(clusterId)) {
+      clusterPasswordProperties.put(clusterId, new ConcurrentHashMap<>());
+    }
+    if (!clusterPasswordProperties.get(clusterId).containsKey(configType)) {
```

Comment:
I think it may be replaced with `java.util.concurrent.ConcurrentHashMap#computeIfAbsent` . As a bonus, we would get atomic check/assigment

Reply:
Thanks for this hint (I've to admit I did not know about this option)